File: stepik-union/src/main/java/org/stepik/plugin/refactoring/rename/AbstractRenamePsiElementProcessor.java
package org.stepik.plugin.refactoring.rename;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.refactoring.rename.RenameDialog;
import com.intellij.refactoring.rename.RenamePsiElementProcessor;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;

import static org.stepik.plugin.utils.ProjectPsiFilesUtils.isNotMovableOrRenameElement;

/**
 * @author meanmail
 */
public abstract class AbstractRenamePsiElementProcessor extends RenamePsiElementProcessor {
    private final Set<Class<? extends PsiElement>> acceptableClasses = new HashSet<>();

    protected void addAcceptableClasses(@NotNull Set<Class<? extends PsiElement>> classes) {
        acceptableClasses.addAll(classes);
    }

    @Override
    public boolean canProcessElement(@NotNull PsiElement element) {
        return isNotMovableOrRenameElement(element, acceptableClasses);
    }

    @Override
    public RenameDialog createRenameDialog(
            Project project, PsiElement element, PsiElement nameSuggestionContext, Editor editor) {
        return new StepikRenameDialog(project, element, nameSuggestionContext, editor);
    }
}

File: stepik-union/src/main/java/org/stepik/plugin/refactoring/rename/AbstractRenamePsiElementProcessor.kt
package org.stepik.plugin.refactoring.rename

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.refactoring.rename.RenameDialog
import com.intellij.refactoring.rename.RenamePsiElementProcessor
import org.stepik.plugin.utils.ProjectPsiFilesUtils.isNotMovableOrRenameElement
import java.util.*


abstract class AbstractRenamePsiElementProcessor : RenamePsiElementProcessor() {
    private val acceptableClasses = HashSet<Class<out PsiElement>>()

    protected fun addAcceptableClasses(classes: Set<Class<out PsiElement>>) {
        acceptableClasses.addAll(classes)
    }

    override fun canProcessElement(element: PsiElement): Boolean {
        return isNotMovableOrRenameElement(element, acceptableClasses)
    }

    override fun createRenameDialog(
            project: Project, element: PsiElement, nameSuggestionContext: PsiElement, editor: Editor): RenameDialog {
        return StepikRenameDialog(project, element, nameSuggestionContext, editor)
    }
}
